BUG fixing module first() and DateOffset

DeaMariaLeon opened the pull request:
- [ X] closes #45908 
- [ X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Also fixes part of #51284 (I have not worked on `last()`) 

~~Edge cases were only fixed for offsets passed as string.~~ 
~~Trying to make it work for `DateOffset` would make it more complicated. Should this be fixed or should `DateOffset` be removed from the function (and docs)?~~
~~Will `DataOffset` be [deprecated](https://github.com/pandas-dev/pandas/issues/48262) or fixed? it has bugs or the documentation is incorrect.~~

As commented by @MarcoGorelli:
* Functions `first()` and `last()` assume that the `Series` or `DataFrame` have a sorted `DatetimeIndex`. Only `last()` has it specified in the documentation. 

* `first()` could be easily done as follows (in case it's deprecated): 
`df[df.index < df.index[0] + DateOffset(days= 15)]`

I also tried  to use `first()` with `relativedelta` but I couldn’t make it work.

Edit:

`first()` doesn’t work for `DateOffset`. With this PR `DateOffset` works, but it now seems that there are inconsistencies when given a string or `MonthEnd`. 
I think that since the PRs to fix https://github.com/pandas-dev/pandas/issues/29623, `first()` with a string doesn’t work as normal offset arithmetic.

DeaMariaLeon committed:
BUG first

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon committed:
BUG first

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -9076,9 +9078,20 @@ def first(self, offset) -> Self:
 
         if len(self.index) == 0:
             return self.copy(deep=False)
+        if type(offset) is not offsets.DateOffset:
```
nit: PEP8 suggests using `isinstance` instead of comparing `type` https://peps.python.org/pep-0008/#programming-recommendations

DeaMariaLeon replied:
Thanks

DeaMariaLeon replied:
We can't use `isinstance` I think, because:

```
(Pdb) offset
<2 * MonthEnds>
(Pdb) type(offset) is offsets.DateOffset
False
(Pdb) isinstance(offset, offsets.DateOffset)
True
```

MarcoGorelli replied:
ah I see, thanks for explaining

MarcoGorelli reviewed (commented):
Thanks for working on this!

I'm not super-familiar with offsets but I left a couple of comments, will get back to this

DeaMariaLeon commented on pandas/core/generic.py:
```diff
@@ -9076,9 +9078,20 @@ def first(self, offset) -> Self:
-        offset = to_offset(offset)
-        if not isinstance(offset, Tick) and offset.is_on_offset(self.index[0]):
+        if type(offset) is offsets.DateOffset and offset.is_on_offset(self.index[0]):
+            end_date = self.index[0] - Timedelta(1, unit="d") + offset
```
Thank you for the review! 
Yes, we should not  hard-code like that, I did it as a draft. I wanted to show how different the behavior of DateOffset is from the "Tick" object. 
The whole `first()` was not implemented for DateOffset, there were no tests either. 
Then there are bugs on it (on DateOffset itself).
In the end, I wanted to know if patching `first()` was correct, if DateOffset will be [deprecated](https://github.com/pandas-dev/pandas/issues/48262) or if it should be fixed.

Some of the issues of DateOffset:
1) DateOffset.is_on_offset is different than the offset done with a string.
2) DateOffset.name doesn't work
3) At some point adding an offset to it, was going backwards! (as in self.index[0] + offset = earlier date)

DeaMariaLeon commented:
> do you want to start by making a docs PR to fix this?

Yes

MarcoGorelli reviewed (commented):
Thinking about this further, would it work to just do
```diff
diff --git a/pandas/core/generic.py b/pandas/core/generic.py
index e16ef3580d..274ab41260 100644
--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -43,6 +43,7 @@ from pandas._libs.tslibs import (
     Timestamp,
     to_offset,
 )
+from pandas._libs.tslibs.offsets import DateOffset
 from pandas._typing import (
     AlignJoin,
     AnyArrayLike,
@@ -9073,6 +9074,13 @@ class NDFrame(PandasObject, indexing.IndexingMixin):
         if len(self.index) == 0:
             return self.copy(deep=False)
 
+        if isinstance(offset, DateOffset):
+            end_date = end = self.index[0] + offset
+            if end_date in self.index:
+                end = self.index.searchsorted(end_date, side="left")
+                return self.iloc[:end]
+            return self.loc[:end]
+
         offset = to_offset(offset)
         if not isinstance(offset, Tick) and offset.is_on_offset(self.index[0]):
             # GH#29623 if first value is end of period, remove offset with n = 1
```
? (I'm showing the diff of `generic.py` compared with `upstream/main`)

DeaMariaLeon commented:
> would it work to just do

No... look:
```
>>> x = pd.Series([1] * 10, index=pd.bdate_range("2010-03-31", periods=10))
>>> result = x.first(pd.DateOffset(months=1))
>>> result
2010-03-31    1
2010-04-01    1
2010-04-02    1
2010-04-05    1
2010-04-06    1
2010-04-07    1
2010-04-08    1
2010-04-09    1
2010-04-12    1
2010-04-13    1
Freq: B, dtype: int64
>>> result2 = x.first("1M")
>>> result2
2010-03-31    1
Freq: B, dtype: int64
```

It doesn't need to start with an "end of the month" to fail. 
The reason is that if offset is a DateOffset:
```
offset.is_on_offset(self.index[0])
True
```
If offset is string:
```
offset.is_on_offset(self.index[0])
False
```

We need to substract the `Timedelta(1, unit= ___)` - we could take the unit from the DateOffset.kwds or transform DateOffset to string, to start with. But I'm not sure if that's a good solution.

MarcoGorelli commented:
Isn't that fine though?

Because `DateOffset(months=1)` is different from `MonthEnd(1)` (and `'M'`, somewhat confusingly, means 'month end' - but there is a PR open to deprecate and rename that)

```python
In [4]: pd.Timestamp.now() + pd.DateOffset(months=1)
Out[4]: Timestamp('2023-05-11 16:26:01.139005')

In [5]: pd.Timestamp.now() + pd.tseries.offsets.MonthEnd(1)
Out[5]: Timestamp('2023-04-30 16:26:14.195128')
```

I think it makes sense that
```
x.first(pd.DateOffset(months=1))
```
would take the rows within one calendar month of the first element (i.e. , whereas `x.first('1M')` would only go until the first month-end

DeaMariaLeon commented:
Oh, OK! 

> but there is a PR open to deprecate and rename that

Does that mean that both (string offset and `DateOffset`) will return the same answer eventually?

Should I add examples with `DateOffset` in the docs, and new tests maybe? 

`last()` will need to be fixed too

I don't need to speak @ the meeting, yaaay!

DeaMariaLeon committed:
BUG fixing func first with DO

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon commented:
> Sorry I was unclear

No reason to apologise. Thanks for explaining. 

I will add examples in the docs on another PR if that's fine.

DeaMariaLeon commented:
Sorry but I can't tell if the failures are related to something in my code.

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon commented:
Would somebody very kind review this "annoying" PR @MarcoGorelli ? :)

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+
+    @pytest.mark.parametrize("start, periods", [("2010-03-31", 1), ("2010-03-30", 2)])
+    def test_first_with_first_day_last_of_m_DO(self, frame_or_series, start, periods):
+        x = frame_or_series([1] * 100, index=bdate_range(start, periods=100))
```
just out of interest, any reason to use `bdate_range` instead of `date_range`? the latter might be easier to read (I had to check whether `2010-03-31` as a business date)

for the title, perhaps
```
test_last_day_of_month_with_date_offset
```
? the file is already `test_first_and_last` so maybe we don't need `first` in the title again (even though some other tests have it 😄 )

DeaMariaLeon replied:
I used bdate_range to do it like the existing/previous tests. Same with the name. 

Changing it now..

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        )
+        tm.assert_equal(result, expected)
+
+    def test_first_with_first_day_end_of_frq_n_greater_one_DO(self, frame_or_series):
```
how about `test_date_offset_multiple_days`?

DeaMariaLeon replied:
My names are ridiculous :-D

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        i = pd.date_range("2018-04-09", periods=4, freq="2D")
+        x = DataFrame({"A": [1, 2, 3, 4]}, index=i)
+        result = x.first(pd.DateOffset(days=3))
+        expected = DataFrame(
+            {"A": [1, 2]}, index=pd.date_range("2018-04-09", periods=2, freq="2D")
+        )
```
for this one, perhaps we could construct the output explicitly? something like
```python
expected = pd.DataFrame({'A':[1,2]}, index=to_datetime(['2018-04-09', '2018-04-11']))
```
?

just because `date_range` also uses offsets, and this would guard against potential bugs in that changing the test (and also because this is a regression test linked to a specific issue, so it's probably best to stay as simple and explicit as possible)

DeaMariaLeon replied:
Thanks for correcting and explaining

DeaMariaLeon replied:
I'm also using `date_range` on all the tests I wrote, I'm thinking I should remove it on all my tests..?
If I avoid `date_range` on `expected`, it seems that I have to eliminate it from the input too, as `tm.assert_equal` fails. Or I have to add the `freq`. Does this look horrible?:
```
expected = (
            pd.DataFrame({"A":[1, 2]},
            index=pd.to_datetime(['2018-04-09', '2018-04-11']))
            ).resample('2D').ffill()
```

DeaMariaLeon replied:
Maybe using `resample` is wrong..  I should just write by hand the whole.
Sorry for the noise

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        )
+        tm.assert_equal(result, expected)
+
+    def test_first_w_DateOffset_other(self):
```
not sure what `other` means here 😄 perhaps `test_date_offset_15_days`, if you need to distinguish it from the test above?

DeaMariaLeon replied:
Oops - thanks.
I was doing "fast tests" and later forgot to clean up.

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        x = DataFrame({"A": Series(np.arange(30), index=i)}, index=i)
+        result = x.first(pd.DateOffset(days=15))
+        i2 = pd.date_range("2018-04-09", periods=8, freq="2D")
+        expected = DataFrame({"A": Series(np.arange(8), index=i2)}, index=i2)
```
would just
```
expected = DataFrame({"A": Series(np.arange(8), index=i2)}, index=i2)
```
work the same? fine to keep the input as in the original post, but for `expected` let's try to be as simple as possible

DeaMariaLeon replied:
Actually, I don't see a difference on line 136 and what is  written on the comment.
I assume that I should do more like:
```
expected = DataFrame({"A": np.arange(8)}, index=i2)
```

MarcoGorelli replied:
yup, that's right!

MarcoGorelli commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        tm.assert_equal(result, expected)
+
+    def test_first_with_fst_day_last_of_m_DO_kwds_months(self, frame_or_series):
+        periods = 10
+        x = frame_or_series([1] * 10, index=bdate_range("2010-03-31", periods=periods))
+        result = x.first(pd.DateOffset(months=1))
+        expected = frame_or_series(
+            [1] * periods, index=bdate_range("2010-03-31", periods=periods)
+        )
```
could you explain what this test tests please? it looks like `expected` is the same as `x`

DeaMariaLeon replied:
Before the changes we did, `first()` wasn't working with months. Or I should say, at some point it wasn't working at all with months. But yeah, the test is dumb.

DeaMariaLeon replied:
Sure, thanks

MarcoGorelli reviewed (changes requested):
Nice one, thanks! We can do a follow-up for `last` as well after

DeaMariaLeon commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -1,13 +1,17 @@
     bdate_range,
 )
+
+# from pandas._libs.tslibs import offsets
```
Sorry for the garbage

DeaMariaLeon commented on pandas/tests/frame/methods/test_first_and_last.py:
```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
+        )
+        tm.assert_equal(result, expected)
+
+    def test_first_w_DateOffset(self):
```
Yes

DeaMariaLeon committed:
Modified as suggestions

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon commented:
> Nice one, thanks! We can do a follow-up for `last` as well after

My pleasure... Yes, we can do `last()` as well.

It's unclear to me why you only commented about using `date_range` on one test, but not the others. I mean, I use it on the other tests as well.
And when I removed `date_range` from line 122 (`def test_first_with_date_offset`) the `assert_equal`  failed because `freq=None` on `expected`. So I removed it also from the input `i`. I’m not sure if that’s OK. 

Before I removed `freq`at the input, I tried adding it to `expected`, but the methods I tried used an offset as well.

MarcoGorelli commented on doc/source/whatsnew/v2.0.1.rst:
```diff
@@ -33,6 +33,7 @@ Bug fixes
 - Bug in :meth:`ArrowDtype.__from_arrow__` not respecting if dtype is explicitly given (:issue:`52533`)
 - Bug in :meth:`DataFrame.max` and related casting different :class:`Timestamp` resolutions always to nanoseconds (:issue:`52524`)
 - Bug in :meth:`Series.describe` not returning :class:`ArrowDtype` with ``pyarrow.float64`` type with numeric data (:issue:`52427`)
+- Fixed bug in :func:`first` when used with a DateOffset (:issue:`45908`)
```
does this close both issues? if so:
```suggestion
- Fixed bug in :func:`first` when used with a :class:`DateOffset` (:issue:`45908`)
```

MarcoGorelli reviewed (approved):
Nice work! Just left a minor comment on the whatsnew note

I'll leave open a bit in case others (@jbrockmendel ? @phofl ?) have comments, else can merge 

---

> It's unclear to me why you only commented about using date_range on one test, but not the others. I mean, I use it on the other tests as well.

Just to not repeat myself :)

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon committed:
Corrected whatsnew

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon commented:
Thanks @MarcoGorelli for the review (and the patience).

jbrockmendel commented on doc/source/whatsnew/v2.0.1.rst:
```diff
@@ -33,6 +33,7 @@ Bug fixes
 - Bug in :meth:`ArrowDtype.__from_arrow__` not respecting if dtype is explicitly given (:issue:`52533`)
 - Bug in :meth:`DataFrame.max` and related casting different :class:`Timestamp` resolutions always to nanoseconds (:issue:`52524`)
 - Bug in :meth:`Series.describe` not returning :class:`ArrowDtype` with ``pyarrow.float64`` type with numeric data (:issue:`52427`)
+- Fixed bug in :func:`first` when used with a :class:`DateOffset` (:issue:`45908`)
```
there is no pd.first.  Maybe `:meth:DataFrame.first`?

DeaMariaLeon replied:
Change done

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -9070,7 +9071,15 @@ def first(self, offset) -> Self:
         if len(self.index) == 0:
             return self.copy(deep=False)
 
+        if isinstance(offset, offsets.DateOffset):
```
doesn't the `offset = to_offset(offset)` below make this check unnecessary?

MarcoGorelli replied:
the issue is that doing this check after `to_offset` would make it impossible to distinguish between `MonthEnd`, as that would also satisfy `isinstance(offset, DateOffset)`

MarcoGorelli replied:
they need to be handled differently because something like `DateOffset(days=15)` would always return `True` for `if not isinstance(offset, Tick) and offset.is_on_offset(self.index[0])`, and so would be shifted backward by `offset.base` (this is why in the linked issues, the result only has a single row for `DateOffset` input)

DeaMariaLeon committed:
fixing first

DeaMariaLeon committed:
added var

DeaMariaLeon committed:
removed parenthesis

DeaMariaLeon committed:
Removed duplicated code

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into newbug

DeaMariaLeon commented on pandas/core/generic.py:
```diff
@@ -9070,7 +9071,15 @@ def first(self, offset) -> Self:
+            if end_date in self.index:
+                end = self.index.searchsorted(end_date, side="left")
+                return self.iloc[:end]
+            return self.loc[:end]
```
Done

DeaMariaLeon commented:
Friendly ping @MarcoGorelli @jbrockmendel

MarcoGorelli reviewed (changes requested):
Wait sorry, I gave a bad review here - this introduces an inconsistency between `first('M')` and `first(MonthEnd())` (which should mean the same thing - reminder that 'M' means 'month end'):
```python
In [1]: ser = pd.Series([1,2,3], index=pd.DatetimeIndex(['2020-01-31', '2020-02-01', '2020-03-01']))

In [2]: ser.first(MonthEnd())
Out[2]: 
2020-01-31    1
2020-02-01    2
dtype: int64

In [3]: ser.first('M')
Out[3]: 
2020-01-31    1
dtype: int64
```

@jbrockmendel is there a way to tell if an offset is `DateOffset` but not `MonthEnd`? I was a bit surprised to see
```python
In [11]: isinstance(MonthEnd(), DateOffset)
Out[11]: True
```
. `type(offset) is DateOffset` would technically work, but I don't see it being done anywhere else

(as an aside, the fact that the tests still pass indicates the relatively poor test coverage that this function had to begin with)

DeaMariaLeon commented:
> this introduces an inconsistency between `first('M')` and `first(MonthEnd())` (which should mean the same thing - reminder that 'M' means 'month end'):

I think the original issue #29623 was not a bug but `first()` (with a string) was modified to match the docstring: “For instance, ‘1M’ will display all the rows having their index within the first month.”

If `first` accepts a `DateOffset` then it should always handle this offset in a way which is consistent with `DateOffset` arithmetic. If the first timestamp in the index is '2020-01-31' then a MonthEnd() offset means going to the following month end:

```
from pandas.tseries.offsets import MonthEnd, Day
from pandas.tseries.frequencies import to_offset
from pandas.tseries.offsets import DateOffset

t = pd.Timestamp('2020-01-31')
t + MonthEnd()
Timestamp('2020-02-29 00:00:00’)
```
Adding for example 15 days means going 15 days forward

```
t + Day(15)
Timestamp('2020-02-15 00:00:00’)
```
Here we go to mid February with a 15-day offset. In this example, the timespan covered by an offset of month end is thus longer than that of an offset of 15 days. It would be completely unexpected for the user if the time period returned by `first` would be the opposite (ie shorter with a MonthEnd offset than with a 15-day offset). 

For example:

```
i = pd.date_range('2020/1/31', periods=31, freq="D")
s = pd.Series(range(len(i)), index=i)

print(s.first(MonthEnd()).tail(5))
2020-02-24    24
2020-02-25    25
2020-02-26    26
2020-02-27    27
2020-02-28    28
Freq: D, dtype: int64

print(s.first(DateOffset(months=1)).tail(5))
2020-02-24    24
2020-02-25    25
2020-02-26    26
2020-02-27    27
2020-02-28    28
Freq: D, dtype: int64

print(s.first(‘1M').tail(5))
2020-01-31    0
Freq: D, dtype: int64

print(s.first(‘15D').tail(5))
2020-02-10    10
2020-02-11    11
2020-02-12    12
2020-02-13    13
2020-02-14    14
Freq: D, dtype: int64
```

So I think the options should be the following :
- have `first()` behave with a DateOffset input with normal offset arithmetic and behave with a string input as it does now (what we have on this PR). But explain the difference in the docstring.
- allow only string inputs to avoid any confusion and any change in the current string input behaviour.
- change the behaviour with a string input so that it behaves consistently with month end semantics.  The docstring could be changed to that effect to :'1M' will display all the rows having their index  within the first month end following the first element in the index

I'm not sure if I have to (friendly) ping you :) @MarcoGorelli

DeaMariaLeon commented on pandas/core/generic.py:
```diff
@@ -9091,16 +9092,25 @@ def first(self, offset) -> Self:
 
         # Tick-like, e.g. 3 weeks
-        if isinstance(offset, Tick) and end_date in self.index:
+        if isinstance(offset, Tick) or input_is_offset and end_date in self.index:
```
I'll change that.

MarcoGorelli commented:
> I think the original issue https://github.com/pandas-dev/pandas/issues/29623 was not a bug but first() (with a string) was modified to match the docstring:

I'm inclined to agree, and am pretty tempted to suggest reverting the PR which changed it

I'll add it to Wednesday's agenda

DeaMariaLeon commented:
I will work on this shortly

DeaMariaLeon commented:
Closing this one because `Series/DataFrame.first()` has been deprecated: https://github.com/pandas-dev/pandas/pull/53419
